keras_cv/models/segmentation/deeplab_v3/deeplab_v3.py

```diff
@@ -0,0 +1,256 @@
+            raise ValueError(
+                "Input shapes for both the backbone and DeepLabV3 cannot be "
+                "`None`. Received: input_shape={input_shape} and "
+                "backbone.input_shape={backbone.input_shape[1:]}"
+            )
+
+        height = input_shape[0]
+        width = input_shape[1]
+
+        feature_map = backbone(inputs)
+        if spatial_pyramid_pooling is None:
+            spatial_pyramid_pooling = SpatialPyramidPooling(
+                dilation_rates=[6, 12, 18]
+            )
+
+        spp_outputs = spatial_pyramid_pooling(feature_map)
+
+        encoder_outputs = keras.layers.UpSampling2D(
+            size=(
+                height // 4 // spp_outputs.shape[1],
+                width // 4 // spp_outputs.shape[2],
+            ),
+            interpolation="bilinear",
+        )(spp_outputs)
+
+        if low_level_feature_pyramid_level is not None:
+            low_level_feature_extractor = keras.Model(
+                inputs=backbone.input,
+                outputs=backbone.get_layer(
+                    backbone.pyramid_level_inputs[
+                        f"P{low_level_feature_pyramid_level}"
+                    ]
+                ).output,
+            )
+            low_level_feature_projector = keras.Sequential(
+                [
+                    keras.layers.Conv2D(
+                        name="low_level_feature_conv",
+                        filters=projection_filters,
+                        kernel_size=1,
+                        padding="same",
+                        use_bias=False,
+                    ),
+                    keras.layers.BatchNormalization(
+                        name="low_level_feature_norm"
+                    ),
+                    keras.layers.ReLU(name="low_level_feature_relu"),
+                ]
+            )
+
+            low_level_features = low_level_feature_extractor(inputs)
+            low_level_projected_features = low_level_feature_projector(
+                low_level_features
+            )
+            combined_encoder_outputs = keras.layers.Concatenate(axis=-1)(
+                [encoder_outputs, low_level_projected_features]
+            )
+        else:
+            combined_encoder_outputs = encoder_outputs
+
+        if segmentation_head is None:
+            segmentation_head = keras.Sequential(
+                [
+                    keras.layers.Conv2D(
+                        name="segmentation_head_conv",
+                        filters=256,
+                        kernel_size=1,
+                        padding="same",
+                        use_bias=False,
+                    ),
+                    keras.layers.BatchNormalization(
+                        name="segmentation_head_norm"
+                    ),
+                    keras.layers.ReLU(name="segmentation_head_relu"),
+                    keras.layers.UpSampling2D(
+                        size=(4, 4), interpolation="bilinear"
```

Comment:
If we're always upsampling by 4x4 here, then we need to always make sure that we're using P2 and P4 for the intermediate_features and the encoder_features.

Meaning we'd do something like this
```
encoder = keras.Model(inputs, backbone.get_layer(backbone.pyramid_level_inputs["P4"]).output)
encoder_outputs = encoder(inputs)
intermediate_feature_extractor = keras.Model(inputs, backbone.get_layer(backbone.pyramid_level_inputs["P2"]).output)
intermediate_features = intermediate_feature_extractor(inputs)
```